arangod/Scheduler/SupervisedScheduler.cpp

```diff
@@ -500,6 +506,44 @@ void SupervisedScheduler::shutdown() {
   }
 }
 
+Result SupervisedScheduler::detachThread() {
+  std::lock_guard<std::mutex> guard(_mutex);
+  // First see if we have already reached the limit:
+  if (_numDetached >= _maxNumberDetachedThreads) {
+    return Result(TRI_ERROR_TOO_MANY_DETACHED_THREADS);
+  }
+
+  // Now we have access to the _workerStates and _detachedWorkerStates
+  // Let's first find ourselves in the _workerStates:
+  uint64_t myNumber = Thread::currentThreadNumber();
+  auto it = _workerStates.begin();
+  while (it != _workerStates.end()) {
+    if ((*it)->_thread->threadNumber() == myNumber) {
+      break;
+    }
+    ++it;
+  }
+  if (it == _workerStates.end()) {
+    return Result(TRI_ERROR_INTERNAL,
+                  "scheduler thread for detaching not found");
+  }
+  std::shared_ptr<WorkerState> state = *it;
+  _workerStates.erase(it);
+  // Since the thread is effectively taken out of the pool, decrease the
+  // number of workers.
+  --_numWorkers;
+  {
+    std::unique_lock<std::mutex> guard(state->_mutex);
+    state->_stop = true;  // We will be stopped after the current task is done
+                          // We know that we are working, so we do not
+                          // have to wake the thread.
+  }
+  ++_metricsThreadsStopped;
+  _detachedWorkerStates.push_back(std::move(state));
```

Comment:
If this `push_back` operation fails because OOM, does it mean we will be in an undefined state, potentially with some resources leaked?
We could avoid this by first reserving some more memory in `_detachedWorkerStates` before making any modifications to the scheduler state, and only performing any state mutations afterwards.

Reply:
I think this is OK, since the thread itself still holds a copy of the `shared_ptr` to its own state. When the current task finishes and the thread asks for more work, it will notice that `_stop` in its own state is set to `true` and will terminate. The only thing which will happen is that `cleanupAbandonedStates` will not see it and will not count down `_numDetached`. Therefore I have now added a `try/catch` statement to ignore the error, but only count the `_numDetached`, if the `push_back` has worked.